zcash_client_backend/src/data_api.rs

```diff
@@ -492,6 +492,9 @@ pub trait WalletRead {
     /// will be interpreted as belonging to that account.
     type AccountId: Copy + Debug + Eq + Hash;
 
+    /// Gets the parameters that went into creating an account (e.g. seed+index or uvk).
+    fn get_account(&self, account_id: Self::AccountId) -> Result<Option<Account>, Self::Error>;
```

Comment:
(Adapted from my previous review of this in #1202.)

Either the documentation for this method isn't accurate, or `Account` is incomplete:
- `Account` omits other parameters that went into creating an account (for seed-based accounts that is at least the seed and the account birthday). It's less inaccurate than what was in (and then removed from) #1202 because it includes the seed fingerprint for seed-based accounts (which commits to the seed directly, rather than indirectly via UFVK), but the return value here does not contain sufficient information to call `WalletWrite::create_account` with in a new wallet to recreate the account state.
- `Account` includes UFVK in the seed-based account case, which is not necessary to create a seed-based account (and makes `WalletRead::get_account` mostly duplicative in that regard with `WalletRead::get_unified_full_viewing_keys`).

The documentation should also say what `Ok(None)` means. I note that some other `WalletRead` APIs use this to mean "account is not known to the wallet", while others return an error in that case. Now that `AccountId` is an associated type, we need to rework all of the `WalletRead` methods that now take `Self::AccountId` to consistently handle this case.
- The caller in a generic context can't generate `Self::AccountId` themselves with the current bounds, so the only sources of account IDs will be either other `WalletRead` methods (in which case `Ok(None)` will never be encountered), or values that were passed into the generic code (in which case the generic code cannot do anything useful with `Ok(None)` except generate an error itself). For that reason I think the `WalletRead` methods should all return an error when the account is unknown, and not `Ok(None)`.

Reply:
I've updated the docs both here and on the `Account` type itself.

> Now that AccountId is an associated type, we need to rework all of the WalletRead methods that now take Self::AccountId to consistently handle this case.

I'm not sure what you mean by this. My PR used to include a bunch of `WalletRead` changes for `AccountId` but that was taken in by #1202. What more are you suggesting I do as part of this change? Maybe the bullet you then mentioned is the explanation. I can work on that.

Reply:
@str4d `Ok(None)` can still retain the same semantics, because associated types aren't fully existentialized; they're ultimately "visible" to the caller and there's nothing that prevents an implementation of `WalletRead` from using say a bare `u32` for the account ID, in which case it would be possible for the caller to provide a value that doesn't correspond to an actual account.

Reply:
The trait doesn't define a concrete error type that we could document an error result for missing accounts with. So I'm inclined to leave this as `Option<Account>`.

Reply:
I'm confused because this PR *removes* the `AccountNotFound(AccountId)` case from `data_api::Error`. I know that we are not guaranteed that `Self::Error` is `data_api::Error`, but surely that's not a problem: it's only required that any given `WalletRead` impl will have a `Self::Error` with a suitable case.

Reply:
This is because the backend crate doesn't contain any code that returns this error. It always comes from the sqlite crate and is called `SqliteClientError::AccountUnknown`.